### PR TITLE
Update README.md

### PR DESCRIPTION
**Pull request by emiloslavsky**

### CODE REVIEW

**Comment by emiloslavsky** [comment]
@kpjensen please review and merge.  There's a couple of questions marked with 'XXX' that need your input.

File: README.md [hunk 1]
```diff
@@ -84,30 +72,55 @@ download methods were (1) `dx cat`, and (2) `cat` from a dxfuse mount point.
...
+Rename does not allow removing the target file or directory because DNAnexus API does not support this functionality.
+XXX moved here from limitations.  Still relevant and correct?
 
 ## File upload and closing
 
-Each file open for writing is allocated a 96MiB write buffer in memory, which is uploaded as a file part when full. Currently dxfuse will upload up to 4 parts in parallel. 
-The last part upload and `file-xxxx/close` DNAx operation is called only when a file descriptor is closed and the `FlushFile` fuse operation is triggered.
+Each dxfuse file open for writing is allocated a 96MiB write buffer in memory, which is uploaded as a DNAnexus file part when full. dxfuse will upload up to 4 parts in parallel.
+
+XXX 4 parts per file or 4 parts across all files backed by dxfuse?
```
Comment (kpjensen) [inline README.md:115]: 4 parts across all files being uploaded.

File: README.md [hunk 2]
```diff
@@ -117,31 +130,32 @@ write(1, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"...,
...
 write(1, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 1024) = -1
 ```
 
-Ignoring the `FlushFile` op creates an edge case for creating empty files. For empty files the empty part upload and `file-xxxx/close` are not called until the `ReleaseFileHandle` fuse op is triggered by the kernel. This is only triggered when all open file descriptors have been closed. The downside of this behavior is that the application which is writing does not wait for this operation to return as it does for regular files closed via `FlushFile`. 
+Ignoring the `FlushFile` op for empty files creates an edge case for creating empty files in dxfuse-mounted folders. For empty files, the empty part upload and `file-xxxx/close` are not called until the `ReleaseFileHandle` fuse operatioin is triggered by the kernel when the last open file descriptor for a given file has been closed. The downside of this behavior is that the dxfuse client application creating the empty file is unable to catch errors that may happen during `file-xxxxx/close` API call as it does for non-empty files closed via `FlushFile` fuse operation triggered by application's call to `close(3)`.
```
Comment (kpjensen) [inline README.md:144]: Typo in "operatioin"

File: README.md [hunk 3]
```diff
@@ -117,31 +130,32 @@ write(1, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"...,
...
+as DNAnexus files left in open state are eventually removed by the abandoned upload daemon.
 
 ## Upload benchmarks
 
-Upload benchmarks are from an AWS m5n.xlarge instance running Ubuntu 18.04 with kernel 5.4.0-1048-aws. 
-`dx` and `dxfuse` benchmark commands were run like so. These benchmarks are not exact because they include the wait time until the uploaded file is transitioned to the `closed` state. 
+Upload benchmarks are from an AWS m5n.xlarge instance running Ubuntu 18.04 with kernel 5.4.0-1048-aws.
+`dx` and `dxfuse` benchmark commands were run like so. These benchmarks are not exact because they include the wait time until the uploaded file is transitioned to the `closed` state.
 
 `time dd if=/dev/zero bs=1M count=$SIZE | dx upload -`
```
Comment (kpjensen) [inline README.md:156]: Missing '--wait' arg `| dx upload --wait`

File: README.md [hunk 4]
```diff
@@ -84,30 +72,55 @@ download methods were (1) `dx cat`, and (2) `cat` from a dxfuse mount point.
...
+All `mkdir` operations via dxfuse are treated as `mkdir -p`. This is because dxfuse does not present the realtime state of the project. Folders can be created outside of dxfuse (or in another dxfuse process), and therefore not be visible to the current running dxfuse. A subsequent `mkdir` --> `project-xxxx/newFolder` would return a 422 error, because dxfuse did not know the directory already exists. This design is due to spark behavior where multiple worker nodes sometimes attempt to create the same output directory.
+
+### rename behavior
+
+Rename does not allow removing the target file or directory because DNAnexus API does not support this functionality.
+XXX moved here from limitations.  Still relevant and correct?
```
Comment (kpjensen) [inline README.md:109]: This is relevant and correct, dxfuse does not support atomic rename if the target exists. 
e.g.
```
$ ls -lht MNT/posix/file*
-r--r--r-- 1 root root 6 Aug 20 21:23 file1
-r--r--r-- 1 root root 6 Aug 20 21:23 file
$ mv MNT/posix/file MNT/posix/file1
mv: cannot move 'file' to 'file1': Operation not permitted
```
Supported in 2 distinct operations
```
$ rm MNT/posix/file1
$ mv MNT/posix/file MNT/posix/file1
```

File: README.md [hunk 5]
```diff
@@ -84,30 +72,55 @@ download methods were (1) `dx cat`, and (2) `cat` from a dxfuse mount point.
...
 
 ## File upload and closing
 
-Each file open for writing is allocated a 96MiB write buffer in memory, which is uploaded as a file part when full. Currently dxfuse will upload up to 4 parts in parallel. 
-The last part upload and `file-xxxx/close` DNAx operation is called only when a file descriptor is closed and the `FlushFile` fuse operation is triggered.
+Each dxfuse file open for writing is allocated a 96MiB write buffer in memory, which is uploaded as a DNAnexus file part when full. dxfuse will upload up to 4 parts in parallel.
+
+XXX 4 parts per file or 4 parts across all files backed by dxfuse?
+
+The upload of the last DNAnexus file part and the call of `file-xxxx/close` DNAnexus API operation are performed by dxfuse only when the OS process that created the OS file descriptor closes the OS file descriptor, triggering `FlushFile` fuse operation.
```
Comment (kpjensen) [inline README.md:118]: We should add a section regarding FlusFile operations ignored from forked processes as detailed in DEVEX-1948. File descriptors must be closed by the original process which opened them.

File: README.md (hunk 1 above)
Comment (emiloslavsky) [reply README.md]: updated. Thanks!

File: README.md (hunk 2 above)
Comment (emiloslavsky) [reply README.md]: updated. Thanks!

File: README.md (hunk 4 above)
Comment (emiloslavsky) [reply README.md]: updated. Thanks!

File: README.md (hunk 5 above)
Comment (emiloslavsky) [reply README.md:127]: I'll leave the current minimal mention in place for now.  We can add a more detailed explanation later

File: README.md (hunk 3 above)
Comment (emiloslavsky) [reply README.md]: added.

**Pull request merged**